lib/pulse_reader/pulse_reader.c

```diff
@@ -0,0 +1,202 @@
+    signal->timer_value = 0;
+    signal->pos = 0;
+
+    pulse_reader_set_timebase(signal, PulseReaderUnit64MHz);
+    pulse_reader_set_bittime(signal, 1);
+
+    signal->dma_config_timer.Direction = LL_DMA_DIRECTION_PERIPH_TO_MEMORY;
+    signal->dma_config_timer.PeriphOrM2MSrcAddress = (uint32_t) & (TIM2->CNT);
+    signal->dma_config_timer.PeriphOrM2MSrcIncMode = LL_DMA_PERIPH_NOINCREMENT;
+    signal->dma_config_timer.PeriphOrM2MSrcDataSize = LL_DMA_PDATAALIGN_WORD;
+    signal->dma_config_timer.MemoryOrM2MDstAddress = (uint32_t)signal->timer_buffer;
+    signal->dma_config_timer.MemoryOrM2MDstIncMode = LL_DMA_MEMORY_INCREMENT;
+    signal->dma_config_timer.MemoryOrM2MDstDataSize = LL_DMA_MDATAALIGN_WORD;
+    signal->dma_config_timer.Mode = LL_DMA_MODE_CIRCULAR;
+    signal->dma_config_timer.PeriphRequest =
+        LL_DMAMUX_REQ_GENERATOR0; /* executes LL_DMA_SetPeriphRequest */
+    signal->dma_config_timer.Priority = LL_DMA_PRIORITY_VERYHIGH;
+
+    signal->dma_config_gpio.Direction = LL_DMA_DIRECTION_PERIPH_TO_MEMORY;
+    signal->dma_config_gpio.PeriphOrM2MSrcIncMode = LL_DMA_PERIPH_NOINCREMENT;
+    signal->dma_config_gpio.PeriphOrM2MSrcDataSize = LL_DMA_PDATAALIGN_WORD;
+    signal->dma_config_gpio.MemoryOrM2MDstIncMode = LL_DMA_MEMORY_INCREMENT;
+    signal->dma_config_gpio.MemoryOrM2MDstDataSize = LL_DMA_MDATAALIGN_WORD;
+    signal->dma_config_gpio.Mode = LL_DMA_MODE_CIRCULAR;
+    signal->dma_config_gpio.PeriphRequest =
+        LL_DMAMUX_REQ_GENERATOR0; /* executes LL_DMA_SetPeriphRequest */
+    signal->dma_config_gpio.Priority = LL_DMA_PRIORITY_VERYHIGH;
+
+    return signal;
+}
+
+void pulse_reader_set_timebase(PulseReader* signal, PulseReaderUnit unit) {
+    switch(unit) {
+    case PulseReaderUnit64MHz:
+        signal->unit_multiplier = 1;
+        signal->unit_divider = 1;
+        break;
+    case PulseReaderUnitPicosecond:
+        signal->unit_multiplier = 15625;
+        signal->unit_divider = 1;
+        break;
+    case PulseReaderUnitNanosecond:
+        signal->unit_multiplier = 15625;
+        signal->unit_divider = 1000;
+        break;
+    case PulseReaderUnitMicrosecond:
+        signal->unit_multiplier = 15625;
+        signal->unit_divider = 1000000;
+        break;
+    }
+}
+
+void pulse_reader_set_bittime(PulseReader* signal, uint32_t bit_time) {
+    signal->bit_time = bit_time;
+}
+
+void pulse_reader_free(PulseReader* signal) {
+    free(signal->timer_buffer);
+    free(signal->gpio_buffer);
+    free(signal);
+}
+
+uint32_t pulse_reader_samples(PulseReader* signal) {
+    uint32_t dma_pos = signal->size - (uint32_t)LL_DMA_GetDataLength(DMA1, signal->dma_channel);
+
+    return ((signal->pos + signal->size) - dma_pos) % signal->size;
+}
+
+void pulse_reader_stop(PulseReader* signal) {
+    LL_DMA_DisableChannel(DMA1, signal->dma_channel);
+    LL_DMA_DisableChannel(DMA1, signal->dma_channel + 1);
+    LL_DMAMUX_DisableRequestGen(NULL, LL_DMAMUX_REQ_GEN_0);
+    LL_TIM_DisableCounter(TIM2);
```

Comment:
Please, reset gpio mode here. 
furi_hal_gpio_init_simple(signal->gpio, GpioModeAnalog);
is enough